platform-hub-api/app/services/costs_report_generator_service.rb

```diff
@@ -0,0 +1,454 @@
+    }
+
+    data[1..-1].map do |(_, account_id, namespace)|
+      [account_id.to_i, namespace]
+    end.uniq.each do |(account_id, namespace_name)|
+      cluster = clusters_by_account_id_map[account_id]
+      namespace = if cluster
+        KubernetesNamespace
+          .by_cluster(cluster)
+          .by_name(namespace_name)
+          .includes(service: :project)
+          .first
+      else
+        nil
+      end
+      service = namespace&.service
+      project = service&.project
+
+      entry = {
+        account_id: account_id,
+        cluster_name: cluster&.name,
+        namespace_id: namespace&.id,
+        namespace_name: namespace_name,
+        project_id: project&.id,
+        project_shortname: project&.shortname,
+        service_id: service&.id,
+        service_name: service&.name
+      }
+
+      (namespace ? results[:mapped] : results[:unmapped]) << entry
+    end
+
+    results
+  end
+
+  def metrics data
+    columns = data[0]
+    # After the first 3 columns we have metric names
+    start_ix = 3
+    columns[start_ix..-1].map.with_index(start_ix) do |s, ix|
+      {
+        name: s,
+        csv_column_index: ix
+      }
+    end
+  end
+
+  def generate_project_bills accounts:, namespaces:, metrics:, billing_data:, metrics_data:, config:
+    # Example output structure:
+    # {
+    #   '<project_id>' => {
+    #     'shortname' => 'foo',
+    #     'name' => 'foooooooo',
+    #     'cost_centre_code' => 'cost_centre_code',
+    #     'totals' => {
+    #       'clusters' => { 'acp-notprod' => 10200.02, 'acp-prod' => 10100.01 },
+    #       'shared_services' => 10101.1
+    #     },
+    #     'services' => {
+    #       '<service_id>' => {
+    #         'name' => 'bar',
+    #         'daily' => {
+    #           '2017-10-01' => {
+    #             'clusters' => { 'acp-notprod' => 200.02, 'acp-prod' => 100.01 },
+    #             'shared_services' => 101.1
+    #           },
+    #           '2017-10-02' => {
+    #             'clusters' => { 'acp-notprod' => 201.02, 'acp-prod' => 101.01 },
+    #             'shared_services' => 102.1
+    #           }
+    #         },
+    #         'totals' => {
+    #           'clusters' => { 'acp-notprod' => 10200.02, 'acp-prod' => 10100.01 },
+    #           'shared_services' => 10101.1
+    #         }
+    #       }
+    #     }
+    #   }
+    # }
+
+    shared_accounts_by_id = {}
+    non_shared_accounts_by_id = {}
+    accounts.each do |a, obj|
+      account_id = a[:account_id]
+      if config[:shared_costs][:clusters].include?(a[:cluster_name])
+        shared_accounts_by_id[account_id] = a
+      else
+        non_shared_accounts_by_id[account_id] = a
+      end
+    end
+
+    namespaces_by_account_and_name = namespaces.each_with_object(HashInitializer[:hash]) do |n, obj|
+      obj[n[:account_id]][n[:namespace_name]] = n
+    end
+
+    bills_per_day_by_account = billing_data[2..-1].each_with_object(HashInitializer[:hash]) do |r, obj|
```

Comment:
Initially, I did "pre-process" the CSV data in to arrays of hashes, etc. But seemed overkill at the time. The issue is (and this relates to your question re: test coverage) we're not yet sure how this code will evolve – it will certainly change a fair bit when we do the "v2" costs (of layering in AWS resource costs, etc.) and I'm keen on keeping everything isolated to this one method/file for now (even if it is a tad messy). Once this matures we'll have a better idea of how to de-compose it into smaller and more understandable chunks.

Reply:
Cool cool. Let's refactor / test that in future iterations.